### PR TITLE
[RFC]ASoC: SOF: acpi: enable PM for ACPI platforms

### PR DESCRIPTION
**Pull request by ranj063**
This patch adds the platform-specific PM callbacks for
the BYT/CHT ACPI platforms. Since the Minnowboards
do not support PM, the SOF device is prevented from
entering runtime suspend by incrementing the ref count
after the topology is loaded.

This doesnt compile yet. Looking for comments.

### CODE REVIEW

File: sound/soc/sof/core.c [hunk 1]
```diff
@@ -16,6 +16,9 @@
 #include "sof-priv.h"
 #include "ops.h"
 
+bool no_PM_support;
+EXPORT_SYMBOL(no_PM_support);
+
```
Comment (dbaluta) [inline sound/soc/sof/core.c]: @ranj063 I'm not a fan of exporting global variables. Maybe we can integrate this inside sof_dev as another op.
Another option would be to prevent entering runtime suspend if runtime_suspend is not present.

Comment (ranj063) [reply sound/soc/sof/core.c]: @dbaluta agree with this. I'll need to think about a more elegant way to do this.

File: sound/soc/sof/sof-acpi-dev.c [hunk 2]
```diff
@@ -117,6 +118,33 @@ static const struct sof_dev_desc sof_acpi_cherrytrail_desc = {
...
 };
 
+static int byt_rt5651_quirk_cb(const struct dmi_system_id *id)
+{
+	/* S3 is not supported for Minnowboards */
+	no_PM_support = true;
+	return 1;
+}
+
+static const struct dmi_system_id byt_rt5651_quirk_table[] = {
```
Comment (plbossart) [inline sound/soc/sof/sof-acpi-dev.c]: rename as MinnowBoard_quirk_table?

This really has nothing to do with rt5651

File: sound/soc/sof/pcm.c [hunk 3]
```diff
@@ -733,11 +733,11 @@ static int sof_pcm_probe(struct snd_soc_component *component)
 	}
 
 	/*
-	 * Some platforms in SOF, ex: BYT, may not have their platform PM
-	 * callbacks set. Increment the usage count so as to
-	 * prevent the device from entering runtime suspend.
+	 * Some boards such as the Minnowboard do not support PM.
+	 * Increment the usage count so as to prevent the device
+	 * from entering runtime suspend.
```
Comment (plbossart) [inline sound/soc/sof/pcm.c]: For MinnowBoards, isn't it suspend-resume that's broken for audio?
if you use rtcwake -m mem -s 3, would audio work when you resume?

If not, then pm_runtime doesn't help.

File: sound/soc/sof/sof-acpi-dev.c [hunk 4]
```diff
@@ -117,6 +118,33 @@ static const struct sof_dev_desc sof_acpi_cherrytrail_desc = {
 	.arch_ops = &sof_xtensa_arch_ops
 };
 
+static int byt_rt5651_quirk_cb(const struct dmi_system_id *id)
+{
+	/* S3 is not supported for Minnowboards */
+	no_PM_support = true;
```
Comment (plbossart) [inline sound/soc/sof/sof-acpi-dev.c]: we have a sof_acpi_debug variable, you could reclaim one bit to set a DISABLE_SUSPEND.

File: sound/soc/sof/intel/byt.c [hunk 5]
```diff
@@ -483,6 +483,34 @@ static int tangier_pci_probe(struct snd_sof_dev *sdev)
 	return ret;
 }
 
+static int byt_suspend(struct snd_sof_dev *sdev)
```
Comment (cujomalainey) [inline sound/soc/sof/intel/byt.c]: This is inside a `#if IS_ENABLED(CONFIG_SND_SOC_SOF_MERRIFIELD)` block which fails to build baytail fi merrifield is not enabled

File: sound/soc/sof/intel/byt.c [hunk 6]
```diff
@@ -483,6 +483,34 @@ static int tangier_pci_probe(struct snd_sof_dev *sdev)
...
+{
+	/* put DSP into reset, set reset vector and stall */
+	snd_sof_dsp_update_bits64(sdev, BYT_DSP_BAR, SHIM_CSR,
+				  SHIM_BYT_CSR_RST | SHIM_BYT_CSR_VECTOR_SEL |
+				  SHIM_BYT_CSR_STALL,
+				  SHIM_BYT_CSR_RST | SHIM_BYT_CSR_VECTOR_SEL |
+				  SHIM_BYT_CSR_STALL);
```
Comment (cujomalainey) [inline sound/soc/sof/intel/byt.c]: missing return

File: sound/soc/sof/intel/byt.c [hunk 7]
```diff
@@ -483,6 +483,34 @@ static int tangier_pci_probe(struct snd_sof_dev *sdev)
...
+}
+
+static int byt_resume(struct snd_sof_dev *sdev)
+{
+	/* enable Interrupt from both sides */
+	snd_sof_dsp_update_bits64(sdev, BYT_DSP_BAR, SHIM_IMRX, 0x3, 0x0);
+	snd_sof_dsp_update_bits64(sdev, BYT_DSP_BAR, SHIM_IMRD, 0x3, 0x0);
```
Comment (cujomalainey) [inline sound/soc/sof/intel/byt.c]: missing return

File: sound/soc/sof/pcm.c (hunk 3 above)
Comment (ranj063) [reply sound/soc/sof/pcm.c]: @plbossart if I remember correctly, on the minnowboards, not holding the ref count to prevent runtime PM somehow ends up in the DSP getting powered off. I might have to test this again though.

**Comment by cujomalainey** [comment]
Looks like firmware was not loaded again after suspend/resume cycle
```
[  102.344607] sof-audio-acpi 808622A8:00: generating page table for 0000000000b3ed3a size 0xffc0 pages 16
[  102.344611] sof-audio-acpi 808622A8:00: stream_tag 0
[  102.344621] sof-audio-acpi 808622A8:00: ipc tx: 0x60010000
[  102.845378] sof-audio-acpi 808622A8:00: error: ipc timed out for 0x60010000 size 108
[  102.845455] sof-audio-acpi 808622A8:00: invalid header size 0x1120c9c. FW oops is bogus
[  102.845475] sof-audio-acpi 808622A8:00: error: unexpected fault 0x00000000 trace 0x00000000
[  102.845495] sof-audio-acpi 808622A8:00: error: ipc host -> DSP: pending no complete no raw 0x00000000
[  102.845513] sof-audio-acpi 808622A8:00: error: mask host: pending no complete no raw 0x00000000
[  102.845529] sof-audio-acpi 808622A8:00: error: ipc DSP -> host: pending no complete no raw 0x00000000
[  102.845546] sof-audio-acpi 808622A8:00: error: mask DSP: pending no complete no raw 0x00000000
[  102.845564] sof-audio-acpi 808622A8:00: error: hw params ipc failed for stream 0
[  102.845586] sof-audio-acpi 808622A8:00: ASoC: 808622A8:00 hw params failed: -110
[  102.845608]  Low Latency: ASoC: hw_params FE failed -110
```

**Comment by cujomalainey** [comment]
Changes pass suspend/resume on cht (Cyan) :)

**Comment by cujomalainey** [comment]
for github
fixes https://github.com/thesofproject/sof/issues/1898

File: sound/soc/sof/pm.c [hunk 8]
```diff
@@ -337,8 +340,11 @@ static int sof_suspend(struct device *dev, bool runtime_suspend)
 	struct snd_sof_dev *sdev = dev_get_drvdata(dev);
 	int ret;
 
-	/* do nothing if dsp suspend callback is not set */
-	if (!sof_ops(sdev)->suspend)
+	/* do nothing if dsp suspend callbacks are not set */
+	if (!runtime_suspend && !sof_ops(sdev)->suspend)
+		return 0;
+
+	if (runtime_suspend && !sof_ops(sdev)->runtime_suspend)
```
Comment (plbossart) [inline sound/soc/sof/pm.c:347]: should be a separate patch, no?

File: sound/soc/sof/sof-acpi-dev.c [hunk 9]
```diff
@@ -127,13 +161,26 @@ static const struct dev_pm_ops sof_acpi_pm = {
...
 	pm_runtime_set_autosuspend_delay(dev, SND_SOF_SUSPEND_DELAY_MS);
 	pm_runtime_use_autosuspend(dev);
+	pm_runtime_set_active(dev);
 	pm_runtime_enable(dev);
+	pm_runtime_mark_last_busy(dev);
+
+#if IS_ENABLED(CONFIG_SND_SOC_SOF_BAYTRAIL)
+	/* set driver data in the quirk table */
+	for (i = 0; i < ARRAY_SIZE(minnowbard_quirk_table); i++)
+		minnowbard_quirk_table[i].driver_data = dev;
```
Comment (plbossart) [inline sound/soc/sof/sof-acpi-dev.c:179]: this is a bit ugly, no? You could just check the quirk, which sets a variable and then you test said variable.

File: sound/soc/sof/intel/byt.c [hunk 10]
```diff
@@ -749,6 +777,10 @@ const struct snd_sof_dsp_ops sof_cht_ops = {
 	/*Firmware loading */
 	.load_firmware	= snd_sof_load_firmware_memcpy,
 
+	/* PM */
+	.suspend = byt_suspend,
+	.resume = byt_resume,
```
Comment (plbossart) [inline sound/soc/sof/intel/byt.c:782]: is this the same routine for basic PM and runtime PM?

File: sound/soc/sof/sof-acpi-dev.c (hunk 9 above)
Comment (plbossart) [reply sound/soc/sof/sof-acpi-dev.c:179]: and you can also use the sof-acpi-debug kernel parameter to override the DMI and also disable PM!

**Comment by ranj063** [comment]
@plbossart @cujomalainey im ooo today due to the power cut. Will address comments tomorrow

File: sound/soc/sof/sof-acpi-dev.c [hunk 11]
```diff
@@ -117,6 +118,39 @@ static const struct sof_dev_desc sof_acpi_cherrytrail_desc = {
...
 
+static int minnowboard_quirk_cb(const struct dmi_system_id *id)
+{
+	struct device *dev = id->driver_data;
+
+	/*
+	 * Increment the usage count so as to prevent the device
+	 * from entering runtime suspend.
+	 */
+	pm_runtime_get_noresume(dev);
```
Comment (plbossart) [inline sound/soc/sof/sof-acpi-dev.c:129]: btw what happens if instead of taking a fake reference you just don't enable pm_runtime?

**Comment by plbossart** [comment]
Doesn't work for me on Asus T100TA (Baytrail). When I close the lid and reopen it, I see this

<details> <summary> dmesg on resume </summary>
[  272.030757] PM: suspend exit
[  272.485201] sof-audio-acpi 80860F28:00: PCM: stream 0 dir 0 DMA position 9120 DAI position 443232
[  272.989019] sof-audio-acpi 80860F28:00: pcm: trigger stream 0 dir 0 cmd 0
[  272.989028] sof-audio-acpi 80860F28:00: ipc tx: 0x60050000: GLB_STREAM_MSG: TRIG_STOP
[  273.493261] sof-audio-acpi 80860F28:00: error: ipc timed out for 0x60050000 size 12
[  273.493283] sof-audio-acpi 80860F28:00: info: preventing DSP entering D3 state to preserve context
[  273.493322] sof-audio-acpi 80860F28:00: invalid header size 0xef6ff5ef. FW oops is bogus
[  273.493339] sof-audio-acpi 80860F28:00: error: unexpected fault 0x00000000 trace 0x00000000
[  273.493355] sof-audio-acpi 80860F28:00: error: ipc host -> DSP: pending no complete no raw 0x00000000
[  273.493369] sof-audio-acpi 80860F28:00: error: mask host: pending no complete no raw 0x00000000
[  273.493383] sof-audio-acpi 80860F28:00: error: ipc DSP -> host: pending no complete no raw 0x00000000
[  273.493396] sof-audio-acpi 80860F28:00: error: mask DSP: pending no complete no raw 0x00000000
[  273.493409] sof-audio-acpi 80860F28:00: error: waking up any trace sleepers
[  273.493425]  Low Latency: ASoC: trigger FE failed -110
[  273.493482] sof-audio-acpi 80860F28:00: pcm: free stream 0 dir 0
[  273.493499] sof-audio-acpi 80860F28:00: ipc tx: 0x60030000: GLB_STREAM_MSG: PCM_FREE
[  273.997307] sof-audio-acpi 80860F28:00: error: ipc timed out for 0x60030000 size 12
[  273.997329] sof-audio-acpi 80860F28:00: info: preventing DSP entering D3 state to preserve context
[  273.997368] sof-audio-acpi 80860F28:00: invalid header size 0xef6ff5ef. FW oops is bogus
[  273.997384] sof-audio-acpi 80860F28:00: error: unexpected fault 0x00000000 trace 0x00000000
[  273.997401] sof-audio-acpi 80860F28:00: error: ipc host -> DSP: pending no complete no raw 0x00000000
[  273.997415] sof-audio-acpi 80860F28:00: error: mask host: pending no complete no raw 0x00000000
[  273.997428] sof-audio-acpi 80860F28:00: error: ipc DSP -> host: pending no complete no raw 0x00000000
[  273.997442] sof-audio-acpi 80860F28:00: error: mask DSP: pending no complete no raw 0x00000000
[  273.997454] sof-audio-acpi 80860F28:00: error: waking up any trace sleepers
[  273.997634] sof-audio-acpi 80860F28:00: pcm: close stream 0 dir 0
</details>

For some reason rtcwake doesn't work, need to look further into this.

**Comment by plbossart** [comment]
The quirks crap out on my Asus T100 TA

````
[   52.622026] general protection fault: 0000 [#1] SMP PTI
[   52.622034] CPU: 1 PID: 157 Comm: kworker/1:2 Not tainted 5.4.0-rc1-test-test+ #142
[   52.622037] Hardware name: ASUSTeK COMPUTER INC. T100TA/T100TA, BIOS T100TA.304 03/14/2014
[   52.622047] Workqueue: events sof_probe_work [snd_sof]
[   52.622057] RIP: 0010:strstr+0x19/0x70
[   52.622061] Code: f8 75 f0 c3 48 89 f8 c3 66 0f 1f 84 00 00 00 00 00 80 3e 00 74 5d 48 89 f1 48 83 c1 01 80 39 00 75 f7 48 89 f8 48 29 f1 74 48 <80> 3f 00 74 41 48 83 c0 01 80 38 00 75 f7 48 29 f8 48 89 c2 48 39
[   52.622064] RSP: 0018:ffffb641802afdc8 EFLAGS: 00010202
[   52.622068] RAX: 00270013795f1010 RBX: ffffffffc05a1370 RCX: 0000000000000010
[   52.622071] RDX: 0000000000000002 RSI: ffffffffc05a1381 RDI: 00270013795f1010
[   52.622073] RBP: 00000000fffffffe R08: 0000000000000041 R09: 0000000000000004
[   52.622076] R10: 00000012245d6a52 R11: 0000000000000001 R12: 0000000000000000
[   52.622079] R13: 0000000000000000 R14: ffffffffc05a1381 R15: 0ffff9864f66ab10
[   52.622082] FS:  0000000000000000(0000) GS:ffff9864f6680000(0000) knlGS:0000000000000000
[   52.622085] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[   52.622088] CR2: 00007f90a84a0088 CR3: 0000000036fa2000 CR4: 00000000001006e0
[   52.622090] Call Trace:
[   52.622101]  dmi_matches+0x74/0xd0
[   52.622108]  dmi_check_system+0x15/0x50
[   52.622114]  sof_probe_work+0x18b/0x1d0 [snd_sof]
[   52.622123]  process_one_work+0x1f5/0x3b0
[   52.622128]  worker_thread+0x28/0x3c0
[   52.622133]  ? process_one_work+0x3b0/0x3b0
[   52.622137]  kthread+0x10d/0x130
[   52.622142]  ? kthread_park+0x80/0x80
[   52.622148]  ret_from_fork+0x35/0x40
[   52.622152] Modules linked in: snd_sof_acpi snd_sof_intel_byt snd_soc_acpi_intel_match snd_sof_intel_bdw snd_sof_intel_ipc snd_sof snd_sof_xtensa_dsp snd_soc_acpi ax88179_178a usbnet snd_soc_rt5670 snd_soc_rt5651 snd_soc_rt5645 snd_soc_rt5640 intel_soc_dts_thermal snd_seq_midi snd_seq_midi_event snd_soc_rl6231 intel_powerclamp snd_rawmidi snd_soc_core snd_pcm snd_seq mei_txe mei snd_seq_device snd_timer snd int3400_thermal processor_thermal_device int3403_thermal intel_soc_dts_iosf int340x_thermal_zone acpi_thermal_rel int3406_thermal soundcore efivarfs mmc_block i915 i2c_algo_bit drm_kms_helper syscopyarea sysfillrect sysimgblt fb_sys_fops xhci_pci drm xhci_hcd sdhci_acpi sdhci
[   52.622217] ---[ end trace e985fde876d178dc ]---

````

**Comment by plbossart** [comment]
@ranj063 you need to apply this to your quirk table
````diff
diff --git a/sound/soc/sof/sof-acpi-dev.c b/sound/soc/sof/sof-acpi-dev.c
index 9eae05aebeca..ab948c1b12b6 100644
--- a/sound/soc/sof/sof-acpi-dev.c
+++ b/sound/soc/sof/sof-acpi-dev.c
@@ -149,6 +149,7 @@ static struct dmi_system_id minnowbard_quirk_table[] = {
 			DMI_MATCH(DMI_PRODUCT_NAME, "Minnowboard Turbot"),
 		},
 	},
+	{}
 };
 
 #endif
````

**Comment by plbossart** [comment]
It's a complete fail on Asus T100 TA. with this one patch, I get an IPC error
````
speaker-test -Dhw:0,0 -c2 -r48000 -t sine

speaker-test 1.1.0

Playback device is hw:0,0
Stream parameters are 48000Hz, S16_LE, 2 channels
Sine wave rate is 440.0000Hz
Rate set to 48000Hz (requested 48000Hz)
Buffer size range from 96 to 16368
Period size range from 48 to 4080
Using max buffer size 16368
Periods = 4
Unable to set hw params for playback: Connection timed out
Setting of hwparams failed: Connection timed out

````

````
[   37.948073] sof-audio-acpi 80860F28:00: ipc tx: 0x60010000: GLB_STREAM_MSG: PCM_PARAMS
[   38.448962] sof-audio-acpi 80860F28:00: error: ipc timed out for 0x60010000 size 108
[   38.448984] sof-audio-acpi 80860F28:00: info: preventing DSP entering D3 state to preserve context
[   38.449022] sof-audio-acpi 80860F28:00: invalid header size 0x7f6ffdef. FW oops is bogus
[   38.449038] sof-audio-acpi 80860F28:00: error: unexpected fault 0x00000000 trace 0x00000000
[   38.449055] sof-audio-acpi 80860F28:00: error: ipc host -> DSP: pending no complete no raw 0x00000000
[   38.449069] sof-audio-acpi 80860F28:00: error: mask host: pending no complete no raw 0x00000000
[   38.449082] sof-audio-acpi 80860F28:00: error: ipc DSP -> host: pending no complete no raw 0x00000000
[   38.449096] sof-audio-acpi 80860F28:00: error: mask DSP: pending no complete no raw 0x00000000
[   38.449108] sof-audio-acpi 80860F28:00: error: waking up any trace sleepers
[   38.449124] sof-audio-acpi 80860F28:00: error: hw params ipc failed for stream 0
[   38.449139] sof-audio-acpi 80860F28:00: ASoC: 80860F28:00 hw params failed: -110
[   38.449157]  Low Latency: ASoC: hw_params FE failed -110
[   38.449187] sof-audio-acpi 80860F28:00: pcm: free stream 0 dir 0
[   38.449590] sof-audio-acpi 80860F28:00: pcm: close stream 0 dir 0
````

So my guess is that for an ACPI device, you must be using some ACPI power helpers when entering D3, or the setup is not correct.
The ACPI device still shows with an 'unsupported' state even though you've enabled pm_runtime. Something's wrong really.

````
root@plb-T00TA:/sys/bus/acpi/devices/80860F28:00# more power/runtime_status 
unsupported
````

**Comment by plbossart** [comment]
on the positive side of things, the legacy baytrail driver (byt-rt5640), not even the Atom/SST stuff, seems to support S3 suspend/resume while doing audio playback, so there's light at the end of the tunnel.

File: sound/soc/sof/sof-acpi-dev.c (hunk 11 above)
Comment (ranj063) [reply sound/soc/sof/sof-acpi-dev.c:129]: @plbossart this is what I want to do ideally. But the one test that fails with this change is the ipc_flood_test(). When this test runs, we call pm_runtime_get_sync() to resume the DSP if it is suspended and this throws an error is runtime_pm is not enabled.
So if I add a check to see if runtime PM is enabled before calling get_sync(), then it would solve the problem. And of course, every other place we add a pm_runtime_get_sync() would need to the same too. What do you think?

**Comment by ranj063** [comment]
> It's a complete fail on Asus T100 TA. with this one patch, I get an IPC error

@plbossart this looks like the DSP is in D3 and is not resumed when playback is started. Could you please send me the whole dmesg log after you close and open the lid?

**Comment by juimonen** [comment]
@ranj063 should this work on byt-mb? what's the procedure to test it, I would try it out...

**Comment by ranj063** [comment]
closing. will be replaced soon

**Comment by cujomalainey** [comment]
@ranj063 can you link the new PR here once it is pushed?

**Comment by ranj063** [comment]
@cujomalainey this is the new PR https://github.com/thesofproject/linux/pull/1523 but it is still WIP as there're some fw changes needed to go with it
